src/preprocess.rkt

```diff
@@ -7,95 +7,79 @@
 (provide find-preprocessing preprocess-pcontext remove-unnecessary-preprocessing)
 
 ;; See https://pavpanchekha.com/blog/symmetric-expressions.html
-(define (find-preprocessing expression context rules)
-  ;; Here `*` means a test identity that *may* be equal to `expression`, and
-  ;; `~` means the simplest form of an expression.
-  (define variables (context-vars context))
-  (define variable-representations (context-var-reprs context))
-  (define evens*
-    (for/list ([variable (in-list variables)]
-               [representation (in-list variable-representations)])
-      ;; TODO: Handle case where neg isn't supported for this representation
-      (define negate (get-parametric-operator 'neg representation))
-      (replace-vars (list (cons variable (list negate variable))) expression)))
-  (define pairs (combinations variables 2))
-  (define swaps*
-    (for/list ([pair (in-list pairs)])
-      (match-define (list a b) pair)
-      (replace-vars (list (cons a b) (cons b a)) expression)))
-  (define query (make-egg-query (cons expression (append evens* swaps*)) rules))
-  (match-define (cons expression~ rest~) (map last (simplify-batch query)))
-  (define-values (evens~ swaps~) (split-at rest~ (length evens*)))
-  (define swaps (filter-map
-                 (lambda (pair swap~) (and (equal? expression~ swap~) pair))
-                 pairs
-                 swaps~))
-  (define components (connected-components variables swaps))
-  (define abs-instructions
-    (for/list ([variable (in-list variables)] [even~ (in-list evens~)]
-               #:when (equal? expression~ even~))
-      (list 'abs variable)))
-  (define sort-instructions
-    (for/list ([component (in-list components)]
-               #:when (> (length component) 1))
-      (cons 'sort component)))
-  ;; Absolute value should happen before sorting
-  (append abs-instructions sort-instructions))
 
-(define (connected-components variables swaps)
-  (define components (disjoint-set (length variables)))
-  (for ([swap (in-list swaps)])
-    (match-define (list a b) swap)
-    (disjoint-set-union! components
-                         (disjoint-set-find! components (index-of variables a))
-                         (disjoint-set-find! components (index-of variables b))))
-  (group-by
-   (compose (curry disjoint-set-find! components) (curry index-of variables))
-   variables))
+;; TODO: Remove
+;; tests : Expression -> Context -> List (Expression, A)
+;; instructions : List A -> Context -> Instruction
+;; operator : Instruction -> (Point -> Point)
+(struct preprocessing-step (tests instructions operator) #:transparent)
```

Comment:
I think this abstraction is hurting way more than it helps. Instead do 3 methods , one per field, with a match in it.